Expose more metrics

tomsisso opened the pull request.

Commit:
expose operation state changes to observers

Commit:
change remove unneeded operation from the signature

Commit:
implement state observer for monitoring

Commit:
add monitoring observer when creating operations in the factories

Commit:
pass metrics collector to the operations factory

Commit:
fix after observer constructor change

Commit:
Change the connection builder to support the new operations factory constructor -
1. Store the chosen Protocol, if was set - instead of creating the operations factory on the set.
2. Create the operations factory only when needed - after build() - to verify that we have the metrics collector (the builder is independent of the order of the sets)

Commit:
fix tests according to the new constructor

Commit:
fix to protected

rami-prilutsky commented on src/main/java/net/spy/memcached/ops/MonitoringOperationStateChangeObserver.java:
```diff
@@ -0,0 +1,32 @@
+        int durationMicros = (int) TimeUnit.NANOSECONDS.toMicros(System.nanoTime() - prevChangeTime);
+        prevChangeTime = System.nanoTime();
+
+        String metricName = "time-from-" + fromState + "-to-" + toState;
```
%s ?

tomsisso replied:
done

rami-prilutsky commented on src/main/java/net/spy/memcached/ops/MonitoringOperationStateChangeObserver.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * An {@link OperationStateChangeObserver} implementation for monitoring purposes.
+ * */
+public class MonitoringOperationStateChangeObserver implements OperationStateChangeObserver {
```
Must we create a new observer instance on each operation? There'll be A LOT of instances in the heap

tomsisso replied:
Nope, changed that.
In short: 
Now there is a single stateless listener - which will get the needed stuff when called, as method params.
+
I've also cached the final metrics names strings on my own - which is ugly because the build is with java 6 - didn't wanna change that or add a dependency on other external module....please take a look.

rami-prilutsky commented on src/main/java/net/spy/memcached/protocol/BaseOperationImpl.java:
```diff
@@ -85,6 +86,7 @@
   public BaseOperationImpl() {
     super();
     creationTime = System.nanoTime();
+    stateObservers = new ArrayList<OperationStateChangeObserver>();
```
ArrayDeque is faster in this case?

tomsisso replied:
why?

rami-prilutsky replied:
Not sure it's still relevant:
https://publicobject.com/2010/07/07/caliper_confirms_reality_linked_list_vs_array_list/

rami-prilutsky commented on src/test/java/net/spy/memcached/ErrorCodeTest.java:
```diff
@@ -39,7 +40,7 @@
   public void testErrorCodes() throws Exception {
     HashMap<Byte, String> errMap = new HashMap<Byte, String>();
-    OperationFactory opFact = new BinaryOperationFactory();
+    OperationFactory opFact = new BinaryOperationFactory(new DummyMetricCollector());
```
I don't see the DummyMetricCollector impl, did u add it?

rami-prilutsky commented on src/main/java/net/spy/memcached/protocol/BaseOperationImpl.java:
```diff
@@ -166,6 +169,19 @@ protected final synchronized void transitionState(OperationState newState) {
+    }
+  }
+
+  public final void addStateObserver(OperationStateChangeObserver observer) {
```
Should synchronize?

tomsisso replied:
currently no, do you want add anyway?

rami-prilutsky replied:
Not sure, when you're going to call addStateObserver?

tomsisso replied:
in the base operations factory, when operation is created, before it is returned to the caller

Commit:
cache metrics names

Commit:
string format instead

Commit:
Create a single monitoring listener instead of per operation:
1. keep last state change time for each operation on the operation itself, and pass it together with the operation when calling to the observer.
2. change state observers signature accordingly.
3. share a single monitoring observer, which is now stateless.
